bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -139,217 +142,123 @@
-    private static final Logger SOURCE_LOG = LoggerFactory.getLogger(Source.class);
-    private static final long MAX_SPLIT_COUNT = 4_000;
-
+  private interface ScanAdvancer<ResultType> {
     /**
-     * Configuration for a Cloud Bigtable connection, a table, and an optional scan.
+     * Get the next unit of work.
      */
-    private final CloudBigtableScanConfiguration configuration;
-    private transient List<SampleRowKeysResponse> sampleRowKeys;
+    ResultType next(ResultScanner resultScanner) throws IOException;
 
     /**
-     * A {@link BoundedSource} for a Cloud Bigtable {@link Table} with a start/stop key range, along
-     * with a potential filter via a {@link Scan}.
+     * is the work complete? Checks for null in the case of {@link Result}, or empty in the case of
+     * an array of Results.
      */
-    protected class SourceWithKeys extends BoundedSource<Result> {
-      private static final long serialVersionUID = 2561066007121040429L;
-
-      /**
-       * Start key for the range to be scanned.
-       */
-      private final byte[] startKey;
-
-      /**
-       * Stop key for the range to be scanned.
-       */
-      private final byte[] stopKey;
-
-      /**
-       * An estimate of the size of the source, in bytes.
-       *
-       * <p>NOTE: This value is a guesstimate. It could be significantly off, especially if there is
-       * a {@link Scan} selected in the configuration. It will also be off if the start and stop
-       * keys are calculated via
-       * {@link CloudBigtableIO.Source#splitIntoBundles(long, PipelineOptions)}.
-       */
-      private final long estimatedSize;
-
-      @VisibleForTesting
-      protected SourceWithKeys(byte[] startKey, byte[] stopKey, long estimatedSize) {
-        if (stopKey.length > 0) {
-          Preconditions.checkState(Bytes.compareTo(startKey, stopKey) < 0,
-              "Source keys not in order: [%s, %s]", Bytes.toStringBinary(startKey),
-              Bytes.toStringBinary(stopKey));
-          Preconditions.checkState(estimatedSize > 0, "Source size must be positive",
-              estimatedSize);
-        }
-        this.startKey = startKey;
-        this.stopKey = stopKey;
-        this.estimatedSize = estimatedSize;
-        SOURCE_LOG.debug("Source with split: {}.", this);
-      }
-
-      /**
-       * Gets an estimate of the size of the source.
-       *
-       * <p>NOTE: This value is a guesstimate. It could be significantly off, especially if there is
-       * a{@link Scan} selected in the configuration. It will also be off if the start and stop keys
-       * are calculated via {@link Source#splitIntoBundles(long, PipelineOptions)}.
-       *
-       * @param options The pipeline options.
-       * @return The estimated size of the source, in bytes.
-       */
-      @Override
-      public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
-        return estimatedSize;
-      }
+    boolean isComplete(ResultType result);
```

Comment:
hasMoreResult()
Returns true if data is read and there could be more data to read, false otherwise.